File: frontend/src/lib/schemas/chunking.ts
import { z } from 'zod';

export const RecursiveChunkingSchema = z.object({
    strategy: z.literal('recursive'),
    max_chunk_size: z.number().int().min(100).max(4000).default(800),
    min_chunk_size: z.number().int().min(10).max(500).default(100),
    overlap: z.number().int().min(0).max(1000).default(150),
    separators: z.array(z.string()).default(['\n\n', '\n', '. ', ' ']),
    keep_separator: z.boolean().default(true),
});

export const SentenceChunkingSchema = z.object({
    strategy: z.literal('sentence'),
    max_sentences_per_chunk: z.number().int().min(1).max(50).default(5),
    min_sentences_per_chunk: z.number().int().min(1).max(10).default(1),
    sentence_overlap: z.number().int().min(0).max(10).default(1),
    language: z.string().default('en'),
    respect_paragraphs: z.boolean().default(true),
});

export const TokenChunkingSchema = z.object({
    strategy: z.literal('token'),
    max_tokens: z.number().int().min(64).max(4096).default(512),
    token_overlap: z.number().int().min(0).max(500).default(50),
    tokenizer: z.enum(['tiktoken', 'sentencepiece', 'hf']).default('tiktoken'),
    truncate_long_tokens: z.boolean().default(false),
    count_special_tokens: z.boolean().default(false),
});

export const SemanticChunkingSchema = z.object({
    strategy: z.literal('semantic'),
    embedding_model: z.string().default('text-embedding-3-small'),
    similarity_threshold: z.number().min(0).max(1).default(0.3),
    max_chunk_tokens: z.number().int().min(100).max(8192).default(1024),
    min_chunk_tokens: z.number().int().min(10).max(1000).default(100),
    merge_small_chunks: z.boolean().default(true),
});

export const FixedLengthChunkingSchema = z.object({
    strategy: z.literal('fixed'),
    chunk_size: z.number().int().min(100).max(10000).default(1000),
    overlap: z.number().int().min(0).max(2000).default(200),
});

export const DocumentStructureChunkingSchema = z.object({
    strategy: z.literal('document'),
    split_by: z.enum(['heading', 'section', 'page']).default('heading'),
    max_section_length: z.number().int().min(500).max(10000).default(2000),
    fallback_strategy: z.enum(['recursive', 'fixed']).default('recursive'),
    preserve_metadata: z.boolean().default(true),
});

export const ChunkingConfigSchema = z.discriminatedUnion('strategy', [
    RecursiveChunkingSchema,
    SentenceChunkingSchema,
    TokenChunkingSchema,
    SemanticChunkingSchema,
    FixedLengthChunkingSchema,
    DocumentStructureChunkingSchema,
]);

export type ChunkingConfig = z.infer<typeof ChunkingConfigSchema>;
export type RecursiveChunkingConfig = z.infer<typeof RecursiveChunkingSchema>;
export type SentenceChunkingConfig = z.infer<typeof SentenceChunkingSchema>;
export type TokenChunkingConfig = z.infer<typeof TokenChunkingSchema>;
export type SemanticChunkingConfig = z.infer<typeof SemanticChunkingSchema>;
export type FixedLengthChunkingConfig = z.infer<typeof FixedLengthChunkingSchema>;
export type DocumentStructureChunkingConfig = z.infer<typeof DocumentStructureChunkingSchema>;

File: backend/app/schemas/chunking.py
from pydantic import BaseModel, Field
from typing import List, Literal, Union, Annotated


class RecursiveChunkingConfig(BaseModel):
    strategy: Literal["recursive"] = "recursive"
    max_chunk_size: int = Field(default=800, ge=10, le=4000)
    min_chunk_size: int = Field(default=100, ge=5, le=500)
    chunk_overlap: int = Field(default=150, ge=0, le=1000)
    separators: List[str] = Field(default=["\n\n", "\n", ". ", " "])
    keep_separator: bool = True
    trim_whitespace: bool = True
    fallback_to_sentence: bool = False


class SentenceChunkingConfig(BaseModel):
    strategy: Literal["sentence"] = "sentence"
    max_sentences_per_chunk: int = Field(default=5, ge=1, le=50)
    min_sentences_per_chunk: int = Field(default=1, ge=1, le=10)
    sentence_overlap: int = Field(default=1, ge=0, le=10)
    language: str = "en"
    respect_paragraphs: bool = True
    merge_short_sentences: bool = True


class TokenChunkingConfig(BaseModel):
    strategy: Literal["token"] = "token"
    max_tokens: int = Field(default=512, ge=2, le=4096)
    token_overlap: int = Field(default=50, ge=0, le=500)
    tokenizer_type: Literal["tiktoken", "sentencepiece", "hf"] = "tiktoken"
    count_special_tokens: bool = False
    truncate_overflow: bool = False
    strict_token_limit: bool = True


class SemanticChunkingConfig(BaseModel):
    strategy: Literal["semantic"] = "semantic"
    embedding_model_ref: str = "text-embedding-3-small"
    similarity_threshold: float = Field(default=0.3, ge=0.0, le=1.0)
    max_chunk_tokens: int = Field(default=1024, ge=100, le=8192)
    min_chunk_tokens: int = Field(default=100, ge=10, le=1000)
    merge_small_chunks: bool = True
    semantic_window_size: int = Field(default=3, ge=1, le=10)


class FixedLengthChunkingConfig(BaseModel):
    strategy: Literal["fixed"] = "fixed"
    chunk_size: int = Field(default=1000, ge=1, le=10000)
    chunk_overlap: int = Field(default=200, ge=0, le=2000)
    hard_cut: bool = False
    pad_last_chunk: bool = False


class DocumentStructureChunkingConfig(BaseModel):
    strategy: Literal["document"] = "document"
    split_by: Literal["heading", "section", "page"] = "heading"
    max_section_length: int = Field(default=2000, ge=500, le=10000)
    fallback_strategy: Literal["recursive", "fixed"] = "recursive"
    preserve_hierarchy: bool = True
    include_metadata: bool = True


ChunkingConfig = Annotated[
    Union[
        RecursiveChunkingConfig,
        SentenceChunkingConfig,
        TokenChunkingConfig,
        SemanticChunkingConfig,
        FixedLengthChunkingConfig,
        DocumentStructureChunkingConfig,
    ],
    Field(discriminator="strategy"),
]
